File: programming-exercises/PHP/second.php
<?php

class NumberSet{
    private $arrayset;

    function __construct($vector = null){
        if ($vector == null){
            $this->arrayset = array();
        }else{
            $this->arrayset = array();
            foreach ($vector as $element) {
      
                if (key_exists($element,$this->arrayset)){
                    throw new Exception('Trying to initalize with a vector with repeated elements');
                }
                $this->arrayset[$element]=True;
            }
        }
    }

    function size(){
        return count($this->arrayset);
    }

    function isEmpty(){
        return count($this->arrayset==0);
    }

    function add($number){
        if (key_exists($number,$this->arrayset)){
            throw new Exception('Trying to add a number which is inside the vector');
        }else{
            $this->arrayset[$number] = True;
        }
    }

    function belong($number){
        return key_exists($number, $this->arrayset);
    }

    function equal(NumberSet $theother){
        return ($this->arrayset == $theother->arrayset );
    }

    function subset(NumberSet $theother){
        foreach (array_keys($theother->arrayset) as $number){
            
            if (!key_exists($number, $this->arrayset)){
                return False;
            }
        }
        return True;
    }

    function vector(){
        return array_keys($this->arrayset);
    }

    function union($theother){
        $union = new NumberSet(array_keys($this->arrayset)); //init $union with this instance numbers 
        foreach (array_keys($theother->arrayset) as $number){ //find which numbers which are in the $theother are not in the $union and add them
            if (!key_exists($number, $union->arrayset)){
                $union->arrayset[$number]=True;
            }
        }
        return $union;
    }
}


echo "hello\n";

$aempty = new NumberSet();

var_dump($aempty);

$first = new NumberSet(array(1,2,3,4));
//~ $fail = new NumberSet(array(1,2,3,3,4)); // throws exception



var_dump($first);

assert($first->size() == 4);
assert($aempty->size() == 0);

assert($aempty->isEmpty());

$aempty->add(1);
assert($aempty->size() == 1);
var_dump($aempty);

$first->add(5);
assert($first->size() == 5);
var_dump($first);

//~ $first->add(3); // THROWS EXCEPTION

assert($first->belong(1));
assert($first->belong(4));
assert($aempty->belong(1));

assert(!$first->equal($aempty));

$equal = new NumberSet(array(5,1,4,3,2));

assert($equal->equal($first));

$second = new NumberSet(array(5,6,7));

assert($equal->subset($first));
assert($first->subset($aempty));
assert(!$aempty->subset($first));
assert(!$first->subset($second));

var_dump($equal->vector());
var_dump($aempty->vector());

var_dump($first->union($aempty));
var_dump($first->union($second));
$third = new NumberSet(array(6,7,8));
var_dump($first->union($third));

var_dump($first->union(new NumberSet()));

echo 'THE END';

?>

File: programming-exercises/Ruby/second.rb
#A hash has been chosen to represent the NumbersSet due to the fact that numbers cannot be repeated inside the set and the hash allow to test easily this feature using the numbers as keys and the values as true or false. 
class NumbersSet
    attr_reader :theNumberSet #define the attributes which are accesible from outside the class.
    
    #Constructor function. It supports 0 arguments and creates an empty NumbersSet or a vector with numbers and creates the numbers set with the numbers inside the vector. It can be created with a Hash too. 
    def initialize(vector = nil)
        @theNumberSet= Hash.new()
        if vector.kind_of?(Array)
            puts 'Initialize: arg Vector'
            vector.each(){|number|
                if @theNumberSet[number]
                    raise '**MY MESSAGE**Vector with repeated component'
                else
                    @theNumberSet[number]=true
                end
            }
        elsif vector.kind_of?(Hash)
            puts 'initialize with a hash'
            @theNumberSet=vector
        else
            puts 'initialize:EmptySet'
        end
    end
    
    def size
        return @theNumberSet.size
    end

    def isEmpty
        return @theNumberSet.empty?
    end

    def add(number)
        if @theNumberSet.include?(number) then
            raise "**MSG** NUMBER REPEATED"
        end
        @theNumberSet[number]=true
    end

    def belong(number)
        return @theNumberSet[number]
    end

    def ==(theOtherSet)
        return @theNumberSet==theOtherSet.theNumberSet
    end

    def <(theSuperSet)
        return !(
                    @theNumberSet.find{|number|
                        !(theSuperSet.belong(number[0]))
                    }
                )
    end

    def to_v()
        @theNumberSet.keys
    end

    def +(theOtherSet)
        theNewNumberSet = @theNumberSet.dup
        theOtherSet.theNumberSet.each_key{|key|
            theNewNumberSet[key]=true
        }
        return NumbersSet.new(theNewNumberSet)
    end
end


aNumberSetEmpty = NumbersSet.new();
#raise "**MSG**NO EMPTY" unless aNumberSetEmpty.isEmpty #EXCEPTION RAISED
aVector = [1,2,3,4]
aRepVector = [1,2,3,4,2]
aNumberSet = NumbersSet.new(aVector)
#aNumberSet = NumbersSet.new(aRepVector) #EXCEPTION RAISED
if aNumberSet.size != 4 then raise "**MSG**SIZE ERROR " end
unless aNumberSetEmpty.isEmpty then raise "**MSG**EMPTY ERROR " end
#aNumberSet.add(2) #EXCEPTION RAISED
aNumberSet.add(9)

unless aNumberSet.belong(9) then raise "**MSG**BELONG ERROR 9 SHOULD BELONG" end
if aNumberSet.belong(11) then raise "**MSG** BELONG ERROR" end

unless aNumberSet==NumbersSet.new([1,2,3,4,9]) then raise "**MSG** WRONG EQUAL, MUST BE TRUE" end

if aNumberSet==NumbersSet.new([1,2,3,4,8]) then raise "**MSG** FAIL IN EQUAL" end
#puts NumbersSet.new([1,2]).belong(2)
if aNumberSet<NumbersSet.new([1,2]) then raise "**MSG** FAIL IN SUBSET" end

unless NumbersSet.new([1,2])<aNumberSet then raise '**MSG** FAIL IN SUBSET' end

puts aNumberSet.to_v().inspect

equalNumberSet = NumbersSet.new([1,2,3,4,9])
almostEqualNumberSet = NumbersSet.new([1,2,3,4,6])
highNumberSet = NumbersSet.new([11,12,13,14])

puts 'UNION BEGINS'
puts (aNumberSet+highNumberSet).to_v().inspect
puts (aNumberSet+almostEqualNumberSet).to_v().inspect
puts (aNumberSet+equalNumberSet).to_v().inspect

puts aNumberSet.inspect();
